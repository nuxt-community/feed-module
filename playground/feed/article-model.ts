export interface Article {
  title: string;
  description: string;
  author_name: string;
  source_name: string;
  source_url: string;
  image_url: string;
  created_at: number;
  inshorts_url: string;
}
