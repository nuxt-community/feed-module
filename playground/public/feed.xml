<?xml version="1.0" encoding="utf-8"?>
<rss version="2.0" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:content="http://purl.org/rss/1.0/modules/content/">
    <channel>
        <title>dupa</title>
        <link>https://www.dupa.pl</link>
        <description>dupa</description>
        <lastBuildDate>Mon, 31 Jan 2022 15:50:06 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>https://github.com/nuxt-modules/feed</generator>
        <copyright>2022 dupa. Ol rajc rizewd</copyright>
        <item>
            <title><![CDATA[Scientists regrow frog's amputated leg in US; pics released]]></title>
            <link>https://www.inshorts.com/en/news/scientists-regrow-frogs-amputated-leg-in-us-pics-released-1643288483296</link>
            <guid>https://www.inshorts.com/en/news/scientists-regrow-frogs-amputated-leg-in-us-pics-released-1643288483296</guid>
            <pubDate>Thu, 27 Jan 2022 05:31:00 GMT</pubDate>
            <content:encoded><![CDATA[A group of American researchers has claimed they have successfully regrown the amputated leg of a frog, using a combination of five drugs. The African clawed frog, which is naturally unable to regenerate its limbs, was treated with the drugs for just 24 hours, which set in motion an 18-month period of regrowth of a functional leg, according to researchers.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Orcas recorded killing and eating blue whales for 1st time in 3 attacks]]></title>
            <link>https://www.inshorts.com/en/news/orcas-recorded-killing-and-eating-blue-whales-for-1st-time-in-3-attacks-1643546516715</link>
            <guid>https://www.inshorts.com/en/news/orcas-recorded-killing-and-eating-blue-whales-for-1st-time-in-3-attacks-1643546516715</guid>
            <pubDate>Sun, 30 Jan 2022 05:11:00 GMT</pubDate>
            <content:encoded><![CDATA[For the first time, orcas, also known as killer whales, have been observed and recorded hunting and killing blue whales, the worldâ€™s largest animal. A study published by Australian researchers explained three instances of a pack of orcas attacking blue whales that included details like a killer whale eating the tongue of a blue whale while it was still alive. 
]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Unknown space object in Milky Way sending radio waves every 18 mins: Astronomers]]></title>
            <link>https://www.inshorts.com/en/news/unknown-space-object-in-milky-way-sending-radio-waves-every-18-mins-astronomers-1643391441675</link>
            <guid>https://www.inshorts.com/en/news/unknown-space-object-in-milky-way-sending-radio-waves-every-18-mins-astronomers-1643391441675</guid>
            <pubDate>Fri, 28 Jan 2022 10:07:00 GMT</pubDate>
            <content:encoded><![CDATA[Researchers from Australia's International Centre for Radio Astronomy Research have said that they've found a "strange" spinning object that releases giant bursts of energy three times an hour or every 18.18 minutes. "It's unlike anything astronomers have seen before," said ICRAR. Located in the Milky Way, the object was found while researchers were mapping radio waves in the Universe.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[IIT Jodhpur develops technique to diagnose COVID using chest X-ray]]></title>
            <link>https://www.inshorts.com/en/news/iit-jodhpur-develops-technique-to-diagnose-covid-using-chest-xray-1643442515539</link>
            <guid>https://www.inshorts.com/en/news/iit-jodhpur-develops-technique-to-diagnose-covid-using-chest-xray-1643442515539</guid>
            <pubDate>Sat, 29 Jan 2022 00:18:00 GMT</pubDate>
            <content:encoded><![CDATA[The researchers at IIT Jodhpur have designed a special AI (Artificial Intelligence) technique to diagnose COVID-19 using chest X-ray. Based on deep learning, they proposed an algorithm called COMiT-Net that studies the abnormalities in lungs to differentiate between normal and COVID-19 infected lungs. They studied 2,500 X-ray images for the research and achieved about 96.8% accuracy.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[SpaceX plans to launch 52 flights in 2022, its highest in a yr]]></title>
            <link>https://www.inshorts.com/en/news/spacex-plans-to-launch-52-flights-in-2022-its-highest-in-a-yr-1643360307658</link>
            <guid>https://www.inshorts.com/en/news/spacex-plans-to-launch-52-flights-in-2022-its-highest-in-a-yr-1643360307658</guid>
            <pubDate>Fri, 28 Jan 2022 01:28:00 GMT</pubDate>
            <content:encoded><![CDATA[Elon Musk-led SpaceX is planning to launch 52 space flights in 2022, a NASA safety panel revealed during a meeting. If SpaceX meets its target, it would be the most launches by it ever conducted in one year, with its previous record last year at 31 launches. SpaceX had aimed 48 launches in 2021 but only made it to 31.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[SpaceX aborts mission 33 secs before launch as ship enters hazard zone]]></title>
            <link>https://www.inshorts.com/en/news/spacex-aborts-mission-33-secs-before-launch-as-ship-enters-hazard-zone-1643642581997</link>
            <guid>https://www.inshorts.com/en/news/spacex-aborts-mission-33-secs-before-launch-as-ship-enters-hazard-zone-1643642581997</guid>
            <pubDate>Mon, 31 Jan 2022 07:53:00 GMT</pubDate>
            <content:encoded><![CDATA[SpaceX aborted a rocket launch 33 seconds before takeoff from Cape Canaveral, Florida, after a cruise liner entered the "hazard zone" on Monday. "We did have a cruise liner making its way towards the no-go zone that the coast guard was unable to clear out in time," SpaceX Production Manager said. SpaceX will try to launch again on February 1.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Carnivory may not have triggered human evolution as believed: US study]]></title>
            <link>https://www.inshorts.com/en/news/carnivory-may-not-have-triggered-human-evolution-as-believed-us-study-1643639308735</link>
            <guid>https://www.inshorts.com/en/news/carnivory-may-not-have-triggered-human-evolution-as-believed-us-study-1643639308735</guid>
            <pubDate>Mon, 31 Jan 2022 06:58:00 GMT</pubDate>
            <content:encoded><![CDATA[US archaeologists have found the evolution of "quintessential" human traits in Homo erectus may not be due to meat consumption, as previously believed. The scientists said that intensive sampling could have ramped up the abundance of modified bones after the Homo erectus appeared. On the other hand, scientists said that meat-eating could have been linked to behavioural and anatomical traits.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[German researchers develop artificial muscles made of natural protein]]></title>
            <link>https://www.inshorts.com/en/news/german-researchers-develop-artificial-muscles-made-of-natural-protein-1643639841832</link>
            <guid>https://www.inshorts.com/en/news/german-researchers-develop-artificial-muscles-made-of-natural-protein-1643639841832</guid>
            <pubDate>Mon, 31 Jan 2022 07:07:00 GMT</pubDate>
            <content:encoded><![CDATA[German researchers developed artificial muscles made from elastin, a natural fibrous protein. The team combined two such proteins, one reacted to temperature and other to pH. The researchers further induced rhythmic contractions in the new material through an external stimulus, sodium sulphite. The study explained that material's contractions could be switched on and off by temperature changes.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[US officials rescue humpback whale from mooring line off Hawaii]]></title>
            <link>https://www.inshorts.com/en/news/us-officials-rescue-humpback-whale-from-mooring-line-off-hawaii-1643565693648</link>
            <guid>https://www.inshorts.com/en/news/us-officials-rescue-humpback-whale-from-mooring-line-off-hawaii-1643565693648</guid>
            <pubDate>Sun, 30 Jan 2022 10:31:00 GMT</pubDate>
            <content:encoded><![CDATA[A rescue team on Wednesday freed a humpback whale from life-threatening entanglement of 140-feet long mooring line and a buoy off Maui, Hawaii, said US' National Oceanic and Atmospheric Administration (NOAA). NOAA added the whale was in good condition but gear around the tail had begun to tear into its flesh. US Coast Guard also assisted NOAA in the operation.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[New species of 'incredibly rare' insect discovered in Uganda]]></title>
            <link>https://www.inshorts.com/en/news/new-species-of-incredibly-rare-insect-discovered-in-uganda-1643369784932</link>
            <guid>https://www.inshorts.com/en/news/new-species-of-incredibly-rare-insect-discovered-in-uganda-1643369784932</guid>
            <pubDate>Fri, 28 Jan 2022 04:06:00 GMT</pubDate>
            <content:encoded><![CDATA[A British scientist has discovered an "incredibly rare" species of insects in Uganda that belongs to a group so rare that its closest relative was last seen in 1969. He has named the metallic-sheened leafhopper Phlogis kibalensis. The 0.26-inch long insect has a pitted body and like most leafhoppers, has uniquely-shaped male reproductive organs, in this case partially leaf-shaped.
]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Cats' brains have shrunk since humans 1st domesticated them: Study]]></title>
            <link>https://www.inshorts.com/en/news/cats-brains-have-shrunk-since-humans-1st-domesticated-them-study-1643370352223</link>
            <guid>https://www.inshorts.com/en/news/cats-brains-have-shrunk-since-humans-1st-domesticated-them-study-1643370352223</guid>
            <pubDate>Fri, 28 Jan 2022 04:15:00 GMT</pubDate>
            <content:encoded><![CDATA[Researchers from the University of Vienna and National Museums Scotland have found that the brain size of domestic cats is smaller than their African ancestors. The decrease in brain size is attributable to domestication. Researchers measured the cranial capacity of large number of domesticated cats, wild African wildcats, European wildcats and a large number of hybrids.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Water flowed on Mars 2 billion years ago, finds NASA]]></title>
            <link>https://www.inshorts.com/en/news/water-flowed-on-mars-2-billion-years-ago-finds-nasa-1643383143240</link>
            <guid>https://www.inshorts.com/en/news/water-flowed-on-mars-2-billion-years-ago-finds-nasa-1643383143240</guid>
            <pubDate>Fri, 28 Jan 2022 07:49:00 GMT</pubDate>
            <content:encoded><![CDATA[NASA's Mars Reconnaissance Orbiter (MRO) has helped researchers to determine that liquid water was present on Mars as old as 2 billion years ago. The researchers studied the chloride salt deposits that were left behind after the ice water flowing across the surface evaporated. The scientists revealed that these salts were found in depressions that once housed shallow ponds.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[COVID-19 booster protects against death from Omicron in over 50s: UK]]></title>
            <link>https://www.inshorts.com/en/news/covid19-booster-protects-against-death-from-omicron-in-over-50s-uk-1643388922950</link>
            <guid>https://www.inshorts.com/en/news/covid19-booster-protects-against-death-from-omicron-in-over-50s-uk-1643388922950</guid>
            <pubDate>Fri, 28 Jan 2022 09:25:00 GMT</pubDate>
            <content:encoded><![CDATA[UK Health Security Agency on Thursday said that protection against death from Omicron in over 50s increased to 95%, two weeks after the administration of COVID-19 booster doses. The authorities added that protection against death from Omicron is 60%, six months after the second dose. Meanwhile, UK reported over 6 lakh cases including over 1,800 deaths in last seven days.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Rare white-clawed crayfish rediscovered in UK nature reserve after 30 years]]></title>
            <link>https://www.inshorts.com/en/news/rare-whiteclawed-crayfish-rediscovered-in-uk-nature-reserve-after-30-years-1643363733173</link>
            <guid>https://www.inshorts.com/en/news/rare-whiteclawed-crayfish-rediscovered-in-uk-nature-reserve-after-30-years-1643363733173</guid>
            <pubDate>Fri, 28 Jan 2022 02:25:00 GMT</pubDate>
            <content:encoded><![CDATA[During routine habitat management, conservationists of Hampshire and Isle of Wight Wildlife Trust rediscovered a cluster of rare, endangered white-clawed crayfish believed to have died out 30 years ago in the Winnall Moors Nature Reserve in Hampshire. A wide chunk of the population of the UK's only native crayfish was thought to have vanished following a deadly plague in 1991.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Chemicals in plastic consumer products may lead to obesity: Study]]></title>
            <link>https://www.inshorts.com/en/news/chemicals-in-plastic-consumer-products-may-lead-to-obesity-study-1643456059664</link>
            <guid>https://www.inshorts.com/en/news/chemicals-in-plastic-consumer-products-may-lead-to-obesity-study-1643456059664</guid>
            <pubDate>Sat, 29 Jan 2022 04:04:00 GMT</pubDate>
            <content:encoded><![CDATA[Researchers at the Norwegian University of Science and Technology studied 34 different plastic products and identified 11 chemical components that interfere with our metabolism, contributing to the development of fat cells. "Our experiments show that ordinary plastic products contain a mix of substances that can be a relevant...factor behind overweight and obesity," Professor Martin Wagner said.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Scotland waste-reduction project aims to recycle hundreds of tonnes of PPE]]></title>
            <link>https://www.inshorts.com/en/news/scotland-wastereduction-project-aims-to-recycle-hundreds-of-tonnes-of-ppe-1643456773092</link>
            <guid>https://www.inshorts.com/en/news/scotland-wastereduction-project-aims-to-recycle-hundreds-of-tonnes-of-ppe-1643456773092</guid>
            <pubDate>Sat, 29 Jan 2022 04:16:00 GMT</pubDate>
            <content:encoded><![CDATA[A research partnership between Scotland's Heriot-Watt University and PPE maker Globus Group aims to recycle hundreds of tonnes of plastic personal protective equipment (PPE) like masks and gloves. Researchers said used plastic PPE will be turned into a secondary raw material called pyrolysis oil, which can then be refined into new commercial products like new PPE products or fuels. ]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Dark matter's origin can be linked to neutrino's origin: IIT Guwahati]]></title>
            <link>https://www.inshorts.com/en/news/dark-matters-origin-can-be-linked-to-neutrinos-origin-iit-guwahati-1643475092896</link>
            <guid>https://www.inshorts.com/en/news/dark-matters-origin-can-be-linked-to-neutrinos-origin-iit-guwahati-1643475092896</guid>
            <pubDate>Sat, 29 Jan 2022 09:21:00 GMT</pubDate>
            <content:encoded><![CDATA[IIT Guwahati researchers revealed that origin and production of dark matter can be related to origin of neutrino mass, one of universe's most abundant particles. The study showed that one of the three lightest right-handed or sterile neutrinos can be the dark matter. As per the Standard Model of particle physics, neutrinos are the only particles with left-handed spin projections.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Air pollution in US rural areas may be as toxic as in urban: Study]]></title>
            <link>https://www.inshorts.com/en/news/air-pollution-in-us-rural-areas-may-be-as-toxic-as-in-urban-study-1643565556849</link>
            <guid>https://www.inshorts.com/en/news/air-pollution-in-us-rural-areas-may-be-as-toxic-as-in-urban-study-1643565556849</guid>
            <pubDate>Sun, 30 Jan 2022 10:29:00 GMT</pubDate>
            <content:encoded><![CDATA[Researchers from University of Illinois Urbana-Champaign found air pollution in US rural areas may be as hazardous as country's urban areas. During the study, scientists found equal pollutant levels in rural and urban areas. However, PM 2.5 levels were found to be more during summer in rural areas suggesting that agriculture activity could be as toxic as an urban setting.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Cooked leafy green vegetables first served 3,500 years ago: Study]]></title>
            <link>https://www.inshorts.com/en/news/cooked-leafy-green-vegetables-first-served-3500-years-ago-study-1643566528831</link>
            <guid>https://www.inshorts.com/en/news/cooked-leafy-green-vegetables-first-served-3500-years-ago-study-1643566528831</guid>
            <pubDate>Sun, 30 Jan 2022 10:45:00 GMT</pubDate>
            <content:encoded><![CDATA[Researchers from Germany, UK said cooked leafy green vegetables were first served 3,500 years ago in West Africa. The scientists studied over 450 prehistoric pots and found lipids (insoluble substances in water) in some of them indicating processed plants of various species. "These plant lipid profiles are most varied...in archaeological pottery to date," said one of the researchers.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Nuclear power risky and expensive to tackle climate crisis, experts say]]></title>
            <link>https://www.inshorts.com/en/news/nuclear-power-risky-and-expensive-to-tackle-climate-crisis-experts-say-1643281020724</link>
            <guid>https://www.inshorts.com/en/news/nuclear-power-risky-and-expensive-to-tackle-climate-crisis-experts-say-1643281020724</guid>
            <pubDate>Thu, 27 Jan 2022 03:27:00 GMT</pubDate>
            <content:encoded><![CDATA[The former heads of nuclear regulatory agencies in the US, France and Germany said nuclear power is too risky and expensive to tackle climate crisis. "Nuclear is just not part of any feasible strategy that could counter climate change. Nuclear is neither clean, safe or smart; but a very complex technology with the potential to cause significant harm," they added.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Record prevalence of COVID-19 in UK in Jan 2022 amid Omicron: Study]]></title>
            <link>https://www.inshorts.com/en/news/record-prevalence-of-covid19-in-uk-in-jan-2022-amid-omicron-study-1643388013368</link>
            <guid>https://www.inshorts.com/en/news/record-prevalence-of-covid19-in-uk-in-jan-2022-amid-omicron-study-1643388013368</guid>
            <pubDate>Fri, 28 Jan 2022 09:10:00 GMT</pubDate>
            <content:encoded><![CDATA[A study by Imperial College London reported record prevalence of COVID-19 in the UK in January 2022 after an Omicron-driven surge in infections. The prevalence between January 5 and January 20 was 4.41%, more than three times higher than it was in December last year. However, researchers said that infections have dropped from the peak and have started plateauing.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[China announces plan to build space centre, satellites for Pak]]></title>
            <link>https://www.inshorts.com/en/news/china-announces-plan-to-build-space-centre-satellites-for-pak-1643391869690</link>
            <guid>https://www.inshorts.com/en/news/china-announces-plan-to-build-space-centre-satellites-for-pak-1643391869690</guid>
            <pubDate>Fri, 28 Jan 2022 10:14:00 GMT</pubDate>
            <content:encoded><![CDATA[China has announced plans to build a space centre and satellites for Pakistan as part of its efforts to boost space cooperation with its ally. Pakistan was mentioned many times in a white paper titled "China's Space Programme: A 2021 Perspective", released by China's State Council. In 2018, China had helped Pakistan to launch two satellites.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[Gas stoves worse for climate due to tiny methane leaks: US study]]></title>
            <link>https://www.inshorts.com/en/news/gas-stoves-worse-for-climate-due-to-tiny-methane-leaks-us-study-1643477310305</link>
            <guid>https://www.inshorts.com/en/news/gas-stoves-worse-for-climate-due-to-tiny-methane-leaks-us-study-1643477310305</guid>
            <pubDate>Sat, 29 Jan 2022 09:58:00 GMT</pubDate>
            <content:encoded><![CDATA[US researchers found gas stoves contributing more to global warming as previously thought due to tiny methane leaks. The scientists said that gas stoves in US are putting 6.8 million tonnes of carbon dioxide into air when they are running. Surprisingly, the scientists also found methane emissions to the tune of 2.6 million tonnes even when the stoves were off. ]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[NASA's JPL gets its first female Director in 85 years]]></title>
            <link>https://www.inshorts.com/en/news/nasas-jpl-gets-its-first-female-director-in-85-years-1643455847834</link>
            <guid>https://www.inshorts.com/en/news/nasas-jpl-gets-its-first-female-director-in-85-years-1643455847834</guid>
            <pubDate>Sat, 29 Jan 2022 04:00:00 GMT</pubDate>
            <content:encoded><![CDATA[Laurie Leshin, President of Worcester Polytechnic Institute (WPI), has become the first woman to be appointed as the Director of NASA's Jet Propulsion Laboratory (JPL) in its 85-year history. She has also been named as the VP of Caltech. She is the first woman President in the university's 150-year history. Faculty and students from Caltech founded JPL in 1936.]]></content:encoded>
        </item>
        <item>
            <title><![CDATA[NASA studies snowstorms in US by flying planes through them]]></title>
            <link>https://www.inshorts.com/en/news/nasa-studies-snowstorms-in-us-by-flying-planes-through-them-1643476051005</link>
            <guid>https://www.inshorts.com/en/news/nasa-studies-snowstorms-in-us-by-flying-planes-through-them-1643476051005</guid>
            <pubDate>Sat, 29 Jan 2022 09:37:00 GMT</pubDate>
            <content:encoded><![CDATA[NASA's Investigation of Microphysics and Precipitation for Atlantic Coast-Threatening Storms (IMPACTS) mission deployed planes equipped with scientific instruments to study US snowstorms. Launched in January, the mission tracks storms by flying two planes at about 65,000 feet, one above the storm and one within clouds. NASA said this will help study snow particles and their precipitation patterns.]]></content:encoded>
        </item>
    </channel>
</rss>