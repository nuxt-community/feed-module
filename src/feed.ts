import { constants } from 'node:fs'
import { access, mkdir, writeFile } from 'node:fs/promises'
import { Feed, FeedOptions } from 'feed'
import { dirname } from 'pathe'
import { joinURL } from 'ufo'
import type { NuxtTemplate } from '@nuxt/schema'
import type { ModuleOptions } from './module'

export interface FeedSource {
  type: 'atom1' | 'json1' | 'rss2'
  path: string
  options?: Partial<FeedOptions>
  create?(options: {
    feed: Feed
    routes: FeedTemplate['options']['routes']
    data: FeedSource['data']
  }): void | Promise<void>
  data?: any[]
}

export type FeedSourcesFactory = () => FeedSource[] | Promise<FeedSource[]>

export async function resolveSources (sources: ModuleOptions['sources']): Promise<FeedSource[]> {
  if (Array.isArray(sources)) {
    return sources
  }

  if (typeof sources === 'function') {
    return await sources()
  }

  throw new TypeError('`sources` option must be an array of feed sources, a function that returns it, or a promise that resolves to it')
}

// cf. NuxtTemplate from @nuxt/schema
export interface FeedTemplate {
  dst: NonNullable<NuxtTemplate['dst']>
  options: {
    source: FeedSource,
    routes: PrerenderGenerateRouteMock[]
  }
}

// Copied PrerenderGenerateRoute from nitropack (not exported)
interface PrerenderGenerateRouteMock {
  route: string
  contents?: string
  data?: ArrayBuffer
  fileName?: string
  error?: Error & {
      statusCode: number
      statusMessage: string
  }
  generateTimeMS?: number
  skip?: boolean
}

// cf. ResolvedNuxtTemplate from @nuxt/schema
export interface ResolvedFeedTemplate {
  dst: string
  contents: string
}

export async function createFeedTemplate (feedTemplate: FeedTemplate): Promise<ResolvedFeedTemplate> {
  const { dst, options: { source, routes } } = feedTemplate
  const { type, path: filePath, create, data, options } = source

  // @ts-expect-error because there are some bugs in upstream type definitons (https://github.com/jpmonette/feed/issues/138).
  // Anyway, let's ignore the error at this moment because they may be configured in `source.create` function.
  const feed = new Feed({
    generator: 'https://github.com/nuxt-community/feed-module',
    ...options,
    link: joinURL(options?.link || '', filePath)
  })

  if (typeof create === 'function') {
    await create({ feed, routes, data })
  }

  // Now it's time to validate feed options.
  // The following functions validate feed options according to their official specification.
  if (type === 'atom1') {
    validateAtomFeedOptions(feed)
  } else if (type === 'json1') {
    validateJsonFeedOptions(feed)
  } else if (type === 'rss2') {
    validateRssFeedOptions(feed)
  }

  const contents = feed[type]()

  return {
    dst,
    contents
  }
}

// Ref. https://www.rfc-editor.org/rfc/rfc4287
function validateAtomFeedOptions (feed: Feed) {
  if (!feed.options.id) {
    throw new Error('Atom feed requires `id` to be configured as a non-empty string in its `source.options` or `source.create` function')
  }

  if (!feed.options.title) {
    throw new Error('Atom feed requires `title` to be configured as a non-empty string in its `source.options` or `source.create` function')
  }

  if (!feed.options.updated) {
    throw new Error('Atom feed requires `updated` to be configured as a non-empty string in its `source.options` or `source.create` function')
  }
}

// Ref. https://www.jsonfeed.org/version/1.1/
function validateJsonFeedOptions (feed: Feed) {
  if (!feed.options.title) {
    throw new Error('JSON feed requires `id` to be configured as a non-empty string in its `source.options` or `source.create` function')
  }
}

// Ref. https://www.rssboard.org/rss-specification
function validateRssFeedOptions (feed: Feed) {
  if (!feed.options.description) {
    throw new Error('RSS feed requires `description` to be configured as a non-empty string in its `source.options` or `source.create` function')
  }

  if (!feed.options.link) {
    throw new Error('RSS feed require `link` to be configured as a non-empty string in their `source.options` or `source.create` function')
  }

  if (!feed.options.title) {
    throw new Error('RSS feed requires `title` to be configured as a non-empty string in its `source.options` or `source.create` function')
  }
}

export async function writeFeedFile (dst: ResolvedFeedTemplate['dst'], data: string): Promise<void> {
  const dir = dirname(dst)

  try {
    await access(dir, constants.W_OK)
  } catch {
    await mkdir(dir, { recursive: true })
  }

  await writeFile(dst, data)
}
