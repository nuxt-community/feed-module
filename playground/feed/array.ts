import axios from 'axios';

import { FeedSource } from '../../src/types';
import { Article } from './article-model';

export const sources: FeedSource[] = [
  {
    meta: {
      id: 'science',
      title: 'Science RSS Feed',
      link: 'https://inshortsv2.vercel.app/news?type=science',
      description: 'Sample RSS feed generated from InShorts v2 API articles',
      copyright: 'Feed module - Maciej PÄ™dzich; InShorts API - Sumit Kolhe'
    },
    path: '/feed.xml',
    type: 'rss2',
    async create(feed) {
      const { data } = await axios.get<{ articles: Article[] }>(
        'https://inshortsv2.vercel.app/news?type=science'
      );
      const { articles } = data;

      for (const article of articles) {
        feed.addItem({
          id: article.created_at.toString(),
          title: article.title,
          link: article.source_url,
          content: article.description,
          date: new Date(article.created_at)
        });
      }
    }
  }
];
