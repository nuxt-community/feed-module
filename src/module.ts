import { defineNuxtModule } from '@nuxt/kit'
import { resolveSources, createFeedTemplate } from './feed'
import type { FeedSource, FeedSourcesFactory } from './feed'

export interface ModuleOptions {
  options?: Partial<FeedSource['options']>
  create?: FeedSource['create']
  data?: FeedSource['data']
  sources: FeedSource[] | FeedSourcesFactory
}

export default defineNuxtModule<ModuleOptions>({
  meta: {
    name: '@nuxtjs/feed',
    configKey: 'feed',
    compatibility: {
      nuxt: '^3.0.0'
    }
  },
  defaults: {
    sources: []
  },
  async setup (options, nuxt) {
    console.log(options)
    const {
      options: commonOptions,
      create: commonCreate,
      data: commonData,
      sources
    } = options
    const _sources = await resolveSources(sources)

    nuxt.hook('build:done', () => {
      const feedTemplates = _sources.map((source) => {
        return createFeedTemplate(nuxt, { source, commonOptions, commonCreate, commonData })
      })
      console.log(feedTemplates)
    })
  }
})
