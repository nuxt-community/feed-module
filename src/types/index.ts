import { Feed, FeedOptions } from 'feed';

export type FeedSource = {
  meta: FeedOptions;
  path: string;
  type: 'rss2' | 'atom1' | 'json1';
  create(feed: Feed): Promise<unknown> | unknown;
};

export type FeedSourcesFactory = () => Promise<FeedSource[]> | FeedSource[];
