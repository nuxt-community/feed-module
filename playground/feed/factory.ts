/* eslint-disable no-console */
import axios from 'axios';

import { FeedConfig, FeedConfigFactory } from '../../src/types';
import { Article } from './article-model';

export const sources: FeedConfigFactory = () => {
  const articleCategories = [
    'national',
    'business',
    'politics',
    'sports',
    'technology',
    'startups',
    'entertainment',
    'education',
    'automobile',
    'science',
    'travel',
    'fashion'
  ];

  return articleCategories.map<FeedConfig>((category) => {
    const apiUrl = `https://inshortsv2.vercel.app/news?type=${category}`;

    return {
      meta: {
        id: category,
        title: `${category}-feed`,
        link: apiUrl,
        description: 'Sample RSS feed generated from InShorts v2 API articles',
        copyright: 'Feed module - Maciej PÄ™dzich; InShorts API - Sumit Kolhe'
      },
      path: `/feeds/${category}.xml`,
      type: 'rss2',
      async create(feed) {
        const { data } = await axios.get<{ articles: Article[] }>(apiUrl);
        const { articles } = data;

        articles.forEach((article) => {
          feed.addItem({
            id: article.created_at.toString(),
            title: article.title,
            link: article.source_url,
            content: article.description,
            date: new Date(article.created_at)
          });
        });
      }
    };
  });
};
