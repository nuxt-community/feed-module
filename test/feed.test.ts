import type { Nuxt } from '@nuxt/schema'
import { expect, it, describe } from 'vitest'
import { resolveSources, createFeedTemplate } from '../src/feed'
import type { FeedSource, FeedSourcesFactory } from '../src/feed'

const nuxtMock = {
  options: {
    rootDir: '/test',
    srcDir: '/test',
    dir: {
      public: 'public'
    }
  }
} as Nuxt

const sources: FeedSource[] = [
  {
    type: 'atom1',
    path: '/feed.atom'
  },
  {
    type: 'json1',
    path: '/feed.json'
  },
  {
    type: 'rss2',
    path: '/feed.xml'
  }
]

const sourcesFactory: FeedSourcesFactory = () => sources

const commonFeedOptions = {
  // This is needed for Atom `updated` and RSS `lastBuildDate` in snapshots not to be changed.
  // Without this defined, the properties are set as the time when feed is created.
  updated: new Date(Date.UTC(2018, 0, 9))
}

describe('feed', () => {
  describe('resolveSources', () => {
    it('should receive an array and is fulfilled with the received array', async () => {
      expect(await resolveSources(sources)).toBe(sources)
    })

    it('should receive a function and is fulfilled with an array return by the function', async () => {
      expect(await resolveSources(sourcesFactory)).toBe(sources)
    })

    it('should throw a type error if neither an array nor a function is passed', () => {
      expect(resolveSources).rejects.toThrowError(/^`sources` option must be an array of feed sources or a function that returns it$/)
    })
  })

  describe('createFeedTemplate', () => {
    it("should returns an object whose `dst` property is a path consited of Nuxt's `public` directory and sources' `path` property", () => {
      const feedTemplate = createFeedTemplate(nuxtMock, { source: sources[0] })
      expect(feedTemplate.dst).toBe('/test/public/feed.atom')
    })

    it('should returns an object whose `getContents` property is a function that returns correct Atom content', async () => {
      const feedTemplate = createFeedTemplate(nuxtMock, { source: sources[0], commonOptions: commonFeedOptions })
      expect(await feedTemplate.getContents()).toMatchSnapshot()
    })

    it('should returns an object whose `getContents` property is a function that returns correct JSON content', async () => {
      const feedTemplate = createFeedTemplate(nuxtMock, { source: sources[1], commonOptions: commonFeedOptions })
      expect(await feedTemplate.getContents()).toMatchSnapshot()
    })

    it('should returns an object whose `getContents` property is a function that returns correct RSS content', async () => {
      const feedTemplate = createFeedTemplate(nuxtMock, { source: sources[2], commonOptions: commonFeedOptions })
      expect(await feedTemplate.getContents()).toMatchSnapshot()
    })
  })
})
