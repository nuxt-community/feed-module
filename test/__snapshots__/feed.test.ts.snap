// Vitest Snapshot v1

exports[`feed > createFeedTemplate > should returns an object whose \`getContents\` property is a function that returns correct Atom content 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<feed xmlns=\\"http://www.w3.org/2005/Atom\\">
    <id/>
    <title/>
    <updated>2018-01-09T00:00:00.000Z</updated>
    <generator>https://github.com/nuxt-community/feed-module</generator>
    <link rel=\\"alternate\\" href=\\"/feed.atom\\"/>
</feed>"
`;

exports[`feed > createFeedTemplate > should returns an object whose \`getContents\` property is a function that returns correct JSON content 1`] = `
"{
    \\"version\\": \\"https://jsonfeed.org/version/1\\",
    \\"home_page_url\\": \\"/feed.json\\",
    \\"items\\": []
}"
`;

exports[`feed > createFeedTemplate > should returns an object whose \`getContents\` property is a function that returns correct RSS content 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title>undefined</title>
        <link>/feed.xml</link>
        <description>undefined</description>
        <lastBuildDate>Tue, 09 Jan 2018 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>https://github.com/nuxt-community/feed-module</generator>
    </channel>
</rss>"
`;
